api/include/opentelemetry/common/spin_lock_mutex.h

```diff
@@ -32,22 +55,42 @@ class SpinLockMutex
    */
   void lock() noexcept
   {
-    /* Note: We expect code protected by this lock to be "fast", i.e. we do NOT incrementally
-     * back-off and wait/notify here, we just loop until we have access, then try again.
-     *
-     * This has the downside that we could be spinning a long time if the exporter is slow.
-     * Note: in C++20x we could use `.wait` to make this slightly better. This should move to
-     * an exponential-back-off / wait strategy.
-     */
-    while (flag_.test_and_set(std::memory_order_acquire))
-      /** TODO - We should immmediately yield if the machine is single processor. */
-      ;
+    for (;;)
+    {
+      // Try once
+      if (!flag_.exchange(true, std::memory_order_acquire))
+      {
+        return;
+      }
+      // Spin-Fast
+      for (std::size_t i = 0; i < SPINLOCK_FAST_ITERATIONS; ++i)
+      {
+        if (try_lock())
+        {
+          return;
+        }
+        // TODO: Issue PAUSE/YIELD instruction to reduce contention.
+        // e.g. __builtin_ia32_pause() / YieldProcessor() / _mm_pause();
+      }
+      // Spin-Yield
+      for (std::size_t i = 0; i < SPINLOCK_YIELD_ITERATIONS; ++i)
+      {
+        if (try_lock())
+        {
+          return;
+        }
+        std::this_thread::yield();
```

Comment:
Would exponential back-off help here to avoid synchronized lock cadence?

Reply:
We could just do a single yield before sleep.   This lock assumes short contention times.


 Any exponential back off would be done via sleep timeout, rather than yield (which is meant to be between the spin-fast delay and an actual sleep delay, in wall clock time)

Reply:
A single yield LGTM.

Reply:
done.